test/functional/client_side_encryption_corpus_tests.js

```diff
@@ -0,0 +1,381 @@
+          return clientEncrypted
+            .db(dataDbName)
+            .collection(dataCollName)
+            .findOne({ _id: corpusCopied._id }, { promoteLongs: false, promoteValues: false });
+        })
+        .then(corpusDecrypted => {
+          expect(toJSONComp(corpusDecrypted)).to.deep.equal(toJSONComp(corpus));
+        })
+        .then(() => {
+          // 8. Load `corpus/corpus_encrypted.json <../corpus/corpus-encrypted.json>`_ to a variable named ``corpus_encrypted_expected``.
+          //    Using ``client`` find the inserted document from ``db.coll`` to a variable named ``corpus_encrypted_actual``.
+
+          //    Iterate over each field of ``corpus_encrypted_expected`` and check the following:
+
+          //    - If the ``algo`` is ``det``, that the value equals the value of the corresponding field in ``corpus_encrypted_actual``.
+          //    - If the ``algo`` is ``rand`` and ``allowed`` is true, that the value does not equal the value of the corresponding field in ``corpus_encrypted_actual``.
+          //    - If ``allowed`` is true, decrypt the value with ``client_encryption``. Decrypt the value of the corresponding field of ``corpus_encrypted`` and validate that they are both equal.
+          //    - If ``allowed`` is false, validate the value exactly equals the value of the corresponding field of ``corpus`` (neither was encrypted).
+          return client
+            .db(dataDbName)
+            .collection(dataCollName)
+            .findOne({ _id: corpusCopied._id }, { promoteLongs: false, promoteValues: false });
+        })
+        .then(corpusEncryptedActual => {
+          return forEachP(Object.keys(corpusEncryptedExpected), key => {
+            return assertion(
+              clientEncryption,
+              key,
+              corpusEncryptedExpected[key],
+              corpusEncryptedActual[key]
+            );
+          });
+        });
+    });
+  }
+
+  // Note: You can uncomment the block below to run the corpus for each individial item
+  // instead of running the entire corpus at once. It is significantly slower,
+  // but gives you higher visibility into why the corpus may be failing
+
+  // function pickValues(obj, key) {
+  //   return {
+  //     _id: obj._id,
+  //     altname_aws: obj.altname_aws,
+  //     altname_local: obj.altname_local,
+  //     [key]: obj[key]
+  //   };
+  // }
+  // Object.keys(corpusAll)
+  //   .filter(x => !copyOverValues.has(x))
+  //   .forEach(key => {
+  //     const corpus = pickValues(corpusAll, key);
+  //     const corpusExpectedEncrypted = pickValues(corpusEncryptedExpectedAll, key);
+
+  //     describe(key, function() {
+  //       defineCorpusTests(corpus, corpusExpectedEncrypted);
+  //       defineCorpusTests(corpus, corpusExpectedEncrypted, true);
+  //     });
+  //   });
+
+  defineCorpusTests(corpusAll, corpusEncryptedExpectedAll);
```

Comment:
nit: can we define all the helper functions in the root level of the test file? It reduces indentation by at least one level for hundreds of lines of code. I also think it improve readability when you see:

```
describe(..., function() {
  defineTests()...
});

function defineTests() {
}
```